Documentation changes (UI and some content)

eoinmackall opened the pull request:
**Preface**
I don't want to insist that any of the changes here are necessary to pull. There are two flavors of changes in this PR (I should have started this PR sooner -- sorry): 1. UI changes to the documentation, 2. small changes (typos, reformatting, addition of one introduction) to the manual. Because I don't know how the UI is handled, and it's very subjective, I'll understand if these changes are rejected (and I can incorporate 2. into a new PR if needed).

**Proposed changes in current PR**
- Reverted to default :html format for documenter.jl (removed dependency on DocumenterVitepress.jl).
_purpose_: homogenize documentation across Oscar.jl, Nemo.jl, AbstractAlgebra.jl, Singular.jl, GAP.jl.
_comments_: I think it looks nice. (I think it also makes it a bit easier for users that want to get started with more than one OSCAR sub-environment-project, like Nemo and Hecke.)
- Added temporary logo/favicon for Hecke.jl.
_purpose_: I don't think these had been made yet.
_comments_: These can both be removed/edited/replaced. If there are preferences or suggestions I can try to implement them (e.g. change in colors, words, images. I made about a dozen variations for the logo, but some of the designs didn't look good with the default sizing, so I went with plaintext).
- Added pages.jl to house the pages array for the makedocs format key.
_purpose_: Controls the layout of the sidebar manually, instead of being autogenerated.
_comments_: I couldn't see how this is implemented in the other projects (e.g. Oscar).
- Moved the How-to guides into a single howto.md file.
_purpose_: The purpose was to differentiate between the "how-to guides" and the "tutorials" in the documentation.
_comments_: I didn't immediately see what was the reason for having the separate (how-to) and (tutorials) sections in the documentation. The goal here is to make this difference more clear (i.e. how-to guides are short and give users immediate tools that they can use in Hecke.jl; tutorials are longer and provide a more detailed walkthrough of the interface functionality).
- Some changes to documentation.
_purpose_: Fix typos, update LaTex formatting, add documentation to algebras file.
_comments_: I should have separated this part out. Will do this in the future.

___

**Example previews of graphical changes**:
*Desktop (Chrome browser)*:
<img width="1600" height="900" alt="Screenshot from 2025-08-20 13-05-08" src="https://github.com/user-attachments/assets/02a9755c-b54c-483c-985d-8061eb00a9d1" />

*Desktop (Zen browser)*: 
<img width="1600" height="900" alt="Screenshot from 2025-08-20 13-03-55" src="https://github.com/user-attachments/assets/659e4a85-dce5-4285-8b3b-6253ab111924" />

*Mobile (iPhone)*:
<img width="1600" height="900" alt="Screenshot from 2025-08-20 13-10-57" src="https://github.com/user-attachments/assets/baa87ec0-f2dd-45d7-989f-922aa966ffed" />

*Mobile (Galaxy Note)*:
<img width="1600" height="900" alt="Screenshot from 2025-08-20 13-11-51" src="https://github.com/user-attachments/assets/562b45a0-2214-4db1-9b05-2f4e2b31d37f" />

thofma commented:
Hi there! First of all, thanks for your PR!

---

> **Preface** I don't want to insist that any of the changes here are necessary to pull. There are two flavors of changes in this PR (I should have started this PR sooner -- sorry): 1. UI changes to the documentation, 2. small changes (typos, reformatting, addition of one introduction) to the manual. Because I don't know how the UI is handled, and it's very subjective, I'll understand if these changes are rejected (and I can incorporate 2. into a new PR if needed).

The changes in 2) look very nice and we definitely want to merge them.

For 1), I am afraid I have to push back on a few of them.

---

> **Proposed changes in current PR**
> 
> * Reverted to default :html format for documenter.jl (removed dependency on DocumenterVitepress.jl).
>   _purpose_: homogenize documentation across Oscar.jl, Nemo.jl, AbstractAlgebra.jl, Singular.jl, GAP.jl.
>   _comments_: I think it looks nice. (I think it also makes it a bit easier for users that want to get started with more than one OSCAR sub-environment-project, like Nemo and Hecke.)

I am not a big fan of the `:html` format. There is in fact a large push towards moving documentation to use Vitepress in the julia ecosystem. For example, https://docs.makie.org/stable/ did the transition (see also https://github.com/search?q=DocumenterVitepress+path%3AProject.toml&type=code). Apart from the (subjective) better looks, the main reason is better organization of the documentation.

One thing that I would like to do is to follow through with the organization of the docs as described in https://diataxis.fr/ (maybe not exactly, but approximately). To quote someone from the julia slack:

> I think the main issue with Julia documentation is that Documenter.jl forces a linear pattern that is not very helpful as documentation grows. I think we need to move all docs to Vitepress and structure them as a tree-like structure with different top level starting points so that different users can find the right docs rather than making an assumption of what is best for everyone
>I don't think you can properly do Diataxis with Documenter.jl because of this

For these reasons, we will stick with Vitepress for now. I expect more packages to follow, since it is getting easier to setup.

---

> * Added temporary logo/favicon for Hecke.jl.
>   _purpose_: I don't think these had been made yet.

Happy with this. We can iterate on the design later.

---

> * Added pages.jl to house the pages array for the makedocs format key.

Not needed as we will stick with Vitepress.

> * Moved the How-to guides into a single howto.md file.
>   _purpose_: The purpose was to differentiate between the "how-to guides" and the "tutorials" in the documentation.
>   _comments_: I didn't immediately see what was the reason for having the separate (how-to) and (tutorials) sections in the documentation. The goal here is to make this difference more clear (i.e. how-to guides are short and give users immediate tools that they can use in Hecke.jl; tutorials are longer and provide a more detailed walkthrough of the interface functionality).

---

They are separated because of the aim to follow through with https://diataxis.fr/.

I noticed that you copied in the ipynb and markdown file for the quaternion tutorial. This is not intended. These things quickly diverge from the actual code and are not really testable. For this reason, we came up with a pipeline, which takes anything in
https://github.com/thofma/Hecke.jl/blob/master/docs/src/tutorials/ and compiles it into a notebook and a corresonding `.md` file and then inserts the stuff into the documentation. This is to make sure, that they always work and are always up to date. (There is a small todo for me here: https://github.com/thofma/Hecke.jl/issues/1819.)

---

> * Some changes to documentation.
>   _purpose_: Fix typos, update LaTex formatting, add documentation to algebras file.
>   _comments_: I should have separated this part out. Will do this in the future.

Very nice! Definitely want this.

---

P.S.: We have been playing around with cyclic algebras lately and are consider adding them as a separate type. Did you already do some work in this direction? Just to make sure that we don't duplicate the effort.

eoinmackall commented:
@thofma 
Hey Tommy,

First, thanks for the feedback!

>The changes in 2) look very nice and we definitely want to merge them. 
For 1), I am afraid I have to push back on a few of them.

For 1), no worries.
For 2), I might move these changes to a new PR (on a clean branch).
___
> I am not a big fan of the :html format. There is in fact a large push towards moving documentation to use Vitepress...

Okay. I can give the :vitepress format more of a shot.

But, let me document some of my criticisms of the :vitepress format as it is right now:
1. The non-linear format felt overwhelming for me (on first encounter). I think that there are two reasons for this: first, independent information is displayed simultaneously (on the home page for example, the boxes "What is Hecke" and "Oscar" are processed roughly in parallel -- on a first encounter -- but they convey different things. Since the viewer isn't making a choice between two equal things (e.g. choosing a youtube video from the recommended), they have to first figure out what the content of both blocks is, and then decide if they want to follow-up with that information. The linear format eliminates processing time, since you are confronted with only one block at once); second, the navigation links "getting started", "tutorials", "how-to guides" all have similar-ish semantic meaning (although I get how they're different now).
2. The animation for transitioning between the home-page layout and either of the "manual", "tutorial", or "how-to" layouts is a big pain point for me -- it feels disorienting (because the location of items in the navigation bar aren't static, e.g. the search bar actually moves places between the two pages).
3. It's different from the format used by the rest of the Oscar ecosystem, so it's something new to learn when switching from one of the other pieces to Hecke.
___

> One thing that I would like to do is to follow through with the organization of the docs as described in..

This makes sense to me. I think that this actually (roughly) aligns with how I was trying to format the docs.

Notably, (in the PR) there are four expandable starting nodes:
1. Hecke.jl -- this contains the introduction, giving a brief explanation about what Hecke.jl is and what it can do. For goal oriented users, it also contains the "How-to guides" which can be used to immediately solve problems. (I think, by-nature, the How-to guides are naturally designed for users _without_ experience using the system, which fits in well with an introduction to the system and its features.);
2. Manual -- this contains the exhaustive list of the structs, functions, their inputs/outputs. This should be a guide (or reference in the diataxis language) to the (or for the) API;
3. Tutorials -- this material is for learning while using Hecke (or for using Hecke while learning something else);
4. For Developers -- this should contain a description of the architecture, an explanation for design decisions for that architecture, and also a description of procedures for how to contribute to the development;
and there is also a final Bibliography for works cited.

I think that this structure benefits a few groups
- researchers probably want to be introduced to the system and get started with their problem,
- students might want to learn while using the system by following a tutorial (maybe researchers too),
- developers will want an explanation for the design of the architecture.
Probably all users will want to see the manual at some point (like looking up a word in a dictionary -- is there a function for doing x?)

I'm not sure if this layout holds up in the :vitepress format.
(Oh I think I just had an epiphany. Maybe the :vitepress format is set-up so that the Home-page is like a window to the program itself. So you can see what it does or what it's like before really reading anything. The other nav-bar links are then for finding targeted information? Maybe I'll tinker around with this...)
___

> Not needed as we will stick with Vitepress.

Actually I was wondering if it was possible to change the names manually for certain pages or files? I really just wanted to change the title "Algebras" to the (what feels a little more descriptive) "Associative Algebras" but couldn't see how to do it.

___
> I noticed that you copied in the ipynb and markdown file for the quaternion tutorial...

Probably this was incidental (to building the documentation locally). I just didn't know that it shouldn't be included (or that it wasn't already there).

___

> P.S.: We have been playing around with cyclic algebras lately and are consider adding them as a separate type. Did you already do some work in this direction? Just to make sure that we don't duplicate the effort.

I don't think I have anything in this direction (although I was thinking about it). I'd be interested! 
(I don't know what was considered but it would be nice if the type could describe a "cyclic algebra" given any cyclic Galois algebra, aside from just cyclic field extensions).

thofma commented:
Let me first address the things that you raise regarding the overall layout:

> 1. The non-linear format felt overwhelming for me (on first encounter). I think that there are two reasons for this: first, independent information is displayed simultaneously (on the home page for example, the boxes "What is Hecke" and "Oscar" are processed roughly in parallel -- on a first encounter -- but they convey different things. Since the viewer isn't making a choice between two equal things (e.g. choosing a youtube video from the recommended), they have to first figure out what the content of both blocks is, and then decide if they want to follow-up with that information. The linear format eliminates processing time, since you are confronted with only one block at once); second, the navigation links "getting started", "tutorials", "how-to guides" all have similar-ish semantic meaning (although I get how they're different now).

> This makes sense to me. I think that this actually (roughly) aligns with how I was trying to format the docs.
> 
> Notably, (in the PR) there are four expandable starting nodes:
> 
> 1. Hecke.jl -- this contains the introduction, giving a brief explanation about what Hecke.jl is and what it can do. For goal oriented users, it also contains the "How-to guides" which can be used to immediately solve problems. (I think, by-nature, the How-to guides are naturally designed for users _without_ experience using the system, which fits in well with an introduction to the system and its features.);
> 2. Manual -- this contains the exhaustive list of the structs, functions, their inputs/outputs. This should be a guide (or reference in the diataxis language) to the (or for the) API;
> 3. Tutorials -- this material is for learning while using Hecke (or for using Hecke while learning something else);
> 4. For Developers -- this should contain a description of the architecture, an explanation for design decisions for that architecture, and also a description of procedures for how to contribute to the development;
>    and there is also a final Bibliography for works cited.
> 
> I think that this structure benefits a few groups
> 
> * researchers probably want to be introduced to the system and get started with their problem,
> * students might want to learn while using the system by following a tutorial (maybe researchers too),
> * developers will want an explanation for the design of the architecture.
>   Probably all users will want to see the manual at some point (like looking up a word in a dictionary -- is there a function for doing x?)
> 
> I'm not sure if this layout holds up in the :vitepress format. (Oh I think I just had an epiphany. Maybe the :vitepress format is set-up so that the Home-page is like a window to the program itself. So you can see what it does or what it's like before really reading anything. The other nav-bar links are then for finding targeted information? Maybe I'll tinker around with this...)

I think the contents of the boxes of the landing page is not great. I just added some fluff to not make it look too empty.

What do you think about making those boxes (which can be made clickable as in https://lux.csail.mit.edu/stable/) links to the different things that you mention:

- Hecke
- Manual
- Tutorials
- For Developers
 
The boxes contains some fluffy text and bring you directly to what you want.

---

> 2. The animation for transitioning between the home-page layout and either of the "manual", "tutorial", or "how-to" layouts is a big pain point for me -- it feels disorienting (because the location of items in the navigation bar aren't static, e.g. the search bar actually moves places between the two pages).

Yeah, it is maybe a bit too much. I can look around and check if we can fix this. It only happens if you switch between pages w/o sidebars. Ideally, there would only be one page without sidebar: "home". So it might not be too bad in practice.

---

> 3. It's different from the format used by the rest of the Oscar ecosystem, so it's something new to learn when switching from one of the other pieces to Hecke.

Yes, I understand this point, but I think the disadvantage does not outweigh the advantages.

---

> Actually I was wondering if it was possible to change the names manually for certain pages or files? I really just wanted to change the title "Algebras" to the (what feels a little more descriptive) "Associative Algebras" but couldn't see how to do it.

Good point. Changing the title is done at the same place, where the file structure is added: https://docs.hecke.thofma.com/stable/manual/developer/documentation#Adding-files-to-the-documentation.

---

> I don't think I have anything in this direction (although I was thinking about it). I'd be interested! (I don't know what was considered but it would be nice if the type could describe a "cyclic algebra" given any cyclic Galois algebra, aside from just cyclic field extensions).

I am having trouble finding a reference for "cyclic Galois algebra". Do you have one at hand?

thofma commented:
> I think everything you suggest sounds great. I can (and would be happy to) write the changes (in this PR) in the coming days. Thank you also for pointing to the documentation on the documentation -- in retrospect I should have checked here first.

Thanks, that would be great. There might be some technical challenges with the automatically created tutorials. I can help with those. Just push your progress to the branch here and I can make adjustments.
 
> > I am having trouble finding a reference for "cyclic Galois algebra". Do you have one at hand?
> 
> My reference is Section 30.A (chapter VII Galois Cohomology) of "The book of involutions" by Knus, Merkurjev, Rost, and Tignol for cyclic algebras defined using Galois algebras (or, Chapter V section 18B for Galois algebras specifically). It was (historically for me) a difficult reference to track down. This reference is for Galois algebras over a field (I might be able to find a more general reference that works for rings, but this was the main case I was thinking about).

Found it, thanks!

eoinmackall commented:
**Changes:**
- Revised Hecke logo (a red quadric surface with some points missing along various conic sections); changed favicon, hero image, and logo to use this image.
_comments_: In Vitepress, the previous logo was a bit too wordy.
- Changed VPHero: added text, tagline, made "features" clickable.
_comments_: After adding the revised logo, I no longer really had any complaints about the placement of features. Now "What is Hecke?" links to the #Features subsection of the homepage, while OSCAR links to the oscar-system website.
- Changed CSS "brand" colorscheme to match Hecke logo; changed CSS Homepage spacing between section headers to fix awkward spacing between Hero and the document container.
_comments_: The current theme is a monochromatic red. I'll include images below.
- Moved "Getting Started" navigation page to website homepage; added some more information in #Getting-Started section; "Getting Started" feature given brand (red) coloring.
_comments_: This was the only navigation page without a file-tree. I spent some time looking at other examples of Vitepress pages, and there was a mix of including this on the homepage or in a new page. At the moment, the "Getting started" section isn't too large, and moving it here eliminated the issue of having multiple navigation links without file trees, so I think this looks like a natural place to put it. I added (a small amount of) content to the #Getting Started section for completeness. (Looking at it now, I think the line that ends with "...session." could end with "...session or script.". I wasn't sure how to frame it, but now it's kind of more for those just starting with Julia.)
- Adjusted file trees
_comments_: I changed title headings in the Manual, separated the Developer documentation from the (user-facing) manual, and moved the references to the end of this section. I added more of the How-to guides into the file tree in "How-to". I also added a file tree to the "Tutorials" page. (I wasn't sure if this was omitted due to the tutorial making pipeline you had mentioned earlier. I wasn't sure how you envisioned this being organized; I think it would -- eventually -- be nice to possibly have the "top-layer" of the manual file tree with tutorials for the respective topics included in each.)
- Moved over edits from previous commit/ made small adjustments (e.g. added a contributing section on the homepage, edited How-to section so that it was reference-able.)
_comments_: This may have been all of the adjustments (together with the Getting Started page).
___
**Dark theme**:
<img width="1600" height="868" alt="s1" src="https://github.com/user-attachments/assets/660f824c-142f-4595-a151-9f8897209161" />
<img width="840" height="904" alt="s2" src="https://github.com/user-attachments/assets/ca46af53-472b-4009-bc17-6c49a901136d" />
<img width="1600" height="900" alt="s3" src="https://github.com/user-attachments/assets/1c12c2b0-59ed-4ace-b248-bc6072b4838d" />

**Light theme**:
<img width="1636" height="904" alt="light1" src="https://github.com/user-attachments/assets/ae16bdf6-46d6-425d-91f8-cbcceb6b99a3" />
<img width="840" height="904" alt="light2" src="https://github.com/user-attachments/assets/0fd6e736-2878-472e-9f28-25e8396bc967" />

thofma commented:
Very nice!

> * Revised Hecke logo (a red quadric surface with some points missing along various conic sections); changed favicon, hero image, and logo to use this image.

Definitely better than what we had before (nothing). Two small comments/questions on the logo thing:
- At the moment the logo looks more geometric than arithmetic. Now I know that logos are hard and I don't want you or me to spend/sink too much time into this. So in case you have ideas about something more arithmetic, I would be happy to hear about this. Otherwise, let's take the surface.
- How do you feel about the red? There was no particular reason back then to pick a red font for the logo in the REPL. It was rather random. So it can also be changed if we want to. (I like that it just works both in dark/light mode.)

Again, these two things are not too imporant.

---

> * Moved "Getting Started" navigation page to website homepage; added some more information in #Getting-Started section; "Getting Started" feature given brand (red) coloring.
>   _comments_: This was the only navigation page without a file-tree. I spent some time looking at other examples of Vitepress pages, and there was a mix of including this on the homepage or in a new page. At the moment, the "Getting started" section isn't too large, and moving it here eliminated the issue of having multiple navigation links without file trees, so I think this looks like a natural place to put it. I added (a small amount of) content to the #Getting Started section for completeness. (Looking at it now, I think the line that ends with "...session." could end with "...session or script.". I wasn't sure how to frame it, but now it's kind of more for those just starting with Julia.)

I like it! In particular I like that cross-references to other sections that you added there (and at other places).

---

> * Adjusted file trees
>   _comments_: I changed title headings in the Manual, separated the Developer documentation from the (user-facing) manual, and moved the references to the end of this section. I added more of the How-to guides into the file tree in "How-to". I also added a file tree to the "Tutorials" page. (I wasn't sure if this was omitted due to the tutorial making pipeline you had mentioned earlier. I wasn't sure how you envisioned this being organized; I think it would -- eventually -- be nice to possibly have the "top-layer" of the manual file tree with tutorials for the respective topics included in each.)

Happy with the changes. Regarding the tutorial organization, do you mean that the the top-layer should be the "topics" instead of the separate tutorials themselves? If so, I agree.

eoinmackall commented:
> At the moment the logo looks more geometric than arithmetic. Now I know that logos are hard and I don't want you or me to spend/sink too much time into this. So in case you have ideas about something more arithmetic, I would be happy to hear about this. Otherwise, let's take the surface.

I agree that it is more geometric. It was difficult (at the time) to think of something arithmetic. I have some ideas now on how arithmetic could be incorporated into the image (e.g. picturing a modular curve, in some capacity). But I'll have to sit down and try to see if these will work out.

_Edit_: I decided to sit-down and try making a couple. I'll attach them below, so there are options. They don't always come out the way I hope they will. If you prefer any of these others, I can switch out the files/favicon/logo.

> How do you feel about the red? There was no particular reason back then to pick a red font for the logo in the REPL. It was rather random. So it can also be changed if we want to. (I like that it just works both in dark/light mode.)

Honestly I'm more or less indifferent to the color. I followed the red color from the REPL, since it was already there. I also like that it seems to work well-enough for both dark/light modes.

> Regarding the tutorial organization, do you mean that the the top-layer should be the "topics" instead of the separate tutorials themselves?

That's right (what I meant), I think. E.g. instead of "Quaternion algebras" and other tutorials, the file structure should be "Tutorials> Number fields/ Orders in Number Fields/ Quadratic and Hermitian Forms/ Associative Algebras / ....". Then, in Associative algebras, you find the Quaternion algebra tutorial.

---
Here are a couple other images. They're all variants on a fundamental domain for X_0(5).
<img width="2048" height="2048" alt="Hecke_logo_modular2" src="https://github.com/user-attachments/assets/d53f5eb4-33f6-4fbf-a430-74e89a397ce8" />
<img width="2048" height="2048" alt="Hecke_logo_shift4" src="https://github.com/user-attachments/assets/efe902b8-fbcc-4c0a-b353-958d421b298a" />
<img width="2048" height="2048" alt="Hecke_logo_modular" src="https://github.com/user-attachments/assets/8c129176-8456-48a2-a534-db7020a6f648" />
